src/pages/Search/components/SearchBar.tsx

```diff
@@ -0,0 +1,49 @@
+import React, { useState } from 'react';
+import { useQuery } from '@tanstack/react-query';
+import SearchBarModal from './SearchBarModal';
+import { getRelateSearch } from '@/api/shop';
+import { RelatedSearchResponse } from '@/api/shop/entity';
+import SearchIconGray from '@/assets/Main/search-icon-gray.svg';
+import useBooleanState from '@/util/hooks/useBooleanState';
+import useDebouncedCallback from '@/util/hooks/useDebounce';
+
+export default function SearchBar() {
+  const [isModalOpen, openModal, closeModal] = useBooleanState(false);
+
+  const [keyword, setKeyword] = useState('');
+
+  const setDebounced = useDebouncedCallback((value: string) => {
+    setKeyword(value.trim());
+  }, 200);
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setDebounced(e.target.value);
+  };
+
+  const { data: relatedSearchItems } = useQuery<RelatedSearchResponse>({
+    queryKey: ['relateSearch', keyword],
+    queryFn: () => getRelateSearch({ keyword }),
+    enabled: keyword.length > 0,
+  });
```

Comment:
1. setBounced와 onChange의 로직은 하나로 묶는 게 자연스러울 것 같아요!
2. 훅을 이벤트 로직 앞에 두는게 일관성 있어 보일 것 같네요!

```suggestion
  const [keyword, setKeyword] = useState('');
  const [isModalOpen, openModal, closeModal] = useBooleanState(false);
  const { data: relatedSearchItems } = useQuery<RelatedSearchResponse>({
    queryKey: ['relateSearch', keyword],
    queryFn: () => getRelateSearch({ keyword }),
    enabled: keyword.length > 0,
  });

  const onChange = useDebouncedCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setKeyword(e.target.value);
  }, 200);
```

3. 그리고 data가 SearchBarModal에서 밖에 쓰이지 않으니 해당 부분으로 데이터 상태를 빼 두는 게 좋겠네요!

```suggestion
  const [keyword, setKeyword] = useState('');
  const [isModalOpen, openModal, closeModal] = useBooleanState(false);

  const onChange = useDebouncedCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setKeyword(e.target.value);
  }, 200);
```

4. debounce 주기는 왜 200으로 설정했는지도 궁금합니다!

Reply:
디바운스와 onChange는 더 간결한 코드로 가능합니다. 또 코드가 잘못된 위치에 있어서 불필요한 렌더링이 발생하기 때문에 코드를 직접 사용하는 파일로 이동하는 게 맞는 것 같습니다. 디바운스 주기는 코인에서 200으로 사용하여 이 곳에서도 200으로 맞췄습니다!